File: Baekjoon/스택/9012 - 괄호.cpp
#include<iostream>
#include<stack>

int main()
{
    int n, flag;
    std::string input;

    scanf("%d", &n);
    for (int i = 0; i < n; i++)
    {
        flag = 0;
        std::stack<char> s;

        std::cin >> input;
        for (int j = 0; j < input.size(); j++)
        {
            if (input[j] == '(')    s.push(input[i]);
            else
            {
                if (s.empty())
                {
                    printf("NO\n");
                    flag++;
                    break;
                }
                else
                {
                    s.pop();
                }
            }
        }
        if (flag) continue;
        if (s.empty())    printf("YES\n");
        else    printf("NO\n");

    }
    system("pause");
}

File: Baekjoon/9012 - 괄호.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;

public class Main {

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringBuilder sb = new StringBuilder();
		int n = Integer.parseInt(br.readLine());
		Stack<Character> stack = new Stack<>();
		String line;
		Boolean flag;
		while (n-- > 0) {
			line = br.readLine();
			flag = false;
			stack.clear();
			for (int j = 0; j < line.length(); j++) {
				if (line.charAt(j) == '(')
					stack.add(line.charAt(j));
				else if (!stack.isEmpty())
					stack.pop();
				else {
					sb.append("NO\n");
					flag = true;
					break;
				}
			}
			if(stack.isEmpty()&&!flag)
				sb.append("YES\n");
			else if (!stack.isEmpty())
				sb.append("NO\n");
		}
		System.out.println(sb);

	}
}
